tensorflow_asr/datasets/asr_dataset.py

```diff
@@ -366,3 +366,165 @@ def create(self, batch_size: int):
         dataset = tf.data.Dataset.from_tensor_slices(self.entries)
         dataset = dataset.map(self.load, num_parallel_calls=AUTOTUNE)
         return self.process(dataset, batch_size)
+
+
+class ASRMaskedSliceDataset(ASRSliceDataset):
+    """ Dataset for ASR with rolling mask """
+
+    def __init__(self,
+                 stage: str,
+                 speech_featurizer: SpeechFeaturizer,
+                 text_featurizer: TextFeaturizer,
+                 data_paths: list,
+                 augmentations: Augmentation = Augmentation(None),
+                 cache: bool = False,
+                 shuffle: bool = False,
+                 indefinite: bool = False,
+                 drop_remainder: bool = True,
+                 use_tf: bool = False,
+                 buffer_size: int = BUFFER_SIZE,
+                 history_window_size: int = 3,
+                 input_chunk_duration: int = 250,
+                 time_reduction_factor: int = 4,
+                 **kwargs):
+        super(ASRMaskedSliceDataset, self).__init__(
+            data_paths=data_paths, augmentations=augmentations,
+            cache=cache, shuffle=shuffle, stage=stage, buffer_size=buffer_size,
+            drop_remainder=drop_remainder, use_tf=use_tf, indefinite=indefinite,
+            speech_featurizer=speech_featurizer, text_featurizer=text_featurizer
+        )
+        self.speech_featurizer = speech_featurizer
+        self.text_featurizer = text_featurizer
+        self.history_window_size = history_window_size
+        self.input_chunk_size = input_chunk_duration * self.speech_featurizer.sample_rate // 1000
+        self.time_reduction_factor = time_reduction_factor
+
+    def calculate_mask(self, num_frames):
```

Comment:
Originally I intended to do a TF only implementation, however I wasn't able to do so.
However, since each sample is re-calculating the same mask only with a different size (since the mask is like a rolling window, it follows the same pattern for all samples, only the size of the mask varies). We could pre-compute a base mask and store it in memory and then sample the original mask to the desired size for each sample.

The only problem I have right now is calculating the original mask with a correct size (max possible size). For this I need to know the maximum audio size in the whole dataset.
I'm also not sure if this would cause some OPs to run on CPU instead of being run on GPU.

Reply:
@andreselizondo-adestech the maximum "feature size" will be calculated in the [metadata.json](https://github.com/TensorSpeech/TensorFlowASR/blob/main/vocabularies/librispeech/librispeech_train_10_1008.metadata.json) from [function](https://github.com/TensorSpeech/TensorFlowASR/blob/6114fb6e69174130779ccaa68430484f6f77999b/tensorflow_asr/datasets/asr_dataset.py#L64). This metadata file is for TPUs support, which is static shape training.

Reply:
Finally done.
I think the implementation is nice. It also includes a mask pre-compute for when metadata is loaded.
Also note, some lines of code could be merged, but I favored readability over performance.
Let me know if there's something you'd like to change. :smile: